tests/component/records/test_hints.py

```diff
@@ -0,0 +1,178 @@
+            required_hints['compression'] = 'BZIP'
+
+            with open(csv_filename, 'rb') as uncompressed_fileobj:
+                gzipped_data = bz2.compress(uncompressed_fileobj.read())
+                fileobj = io.BytesIO(gzipped_data)
+                try:
+                    hints = sniff_hints(fileobj, initial_hints=initial_hints)
+                    self.assertTrue(set(required_hints.items()).issubset(set(hints.items())),
+                                    f"Expected at least these hints while reading {basename}: "
+                                    f"{required_hints}, found these hints: {hints}")
+                except Exception as e:
+                    if str(e) != config.get('raises'):
+                        raise
+
+    def test_sniff_hints_from_fileobjs_nonseekable(self):
+        csv = 'Liberté,égalité,fraternité\n"“a”",2,3\n'
+        csv_bytes = csv.encode('utf-8', errors='replace')
+        with io.BytesIO(csv_bytes) as fileobj:
+            fileobj.seekable = lambda: False
+            with self.assertRaises(OSError):
+                sniff_encoding_hint(fileobj=fileobj)
+
+    def test_sniff_hints_from_fileobjs_encodings(self):
+        expected_hint = {
+            'utf-8': {'hint': 'UTF8'},
+            'utf-8-sig': {'hint': 'UTF8BOM'},
+            # chardet not smart enough right now to sniff UTF16 without BOM:
+            # https://github.com/chardet/chardet/pull/109/files
+            'utf-16': {'hint': 'UTF16', 'initial': 'UTF16'},
+            'utf-16-le': {'hint': 'UTF16LE', 'initial': 'UTF16LE'},
+            'utf-16-be': {'hint': 'UTF16BE', 'initial': 'UTF16BE'},
+            'latin-1': {'hint': 'LATIN1'},
+            'cp1252': {'hint': 'CP1252'},
+        }
+        for python_encoding, test_details in expected_hint.items():
+            csv = 'Liberté,égalité,fraternité\n"“a”",2,3\n'
+            csv_bytes = csv.encode(python_encoding, errors='replace')
+            with io.BytesIO(csv_bytes) as fileobj:
+                fileobjs = [fileobj]
+                initial_hints: PartialRecordsHints = {
+                    'field-delimiter': ','
+                }
+                if 'initial' in test_details:
+                    # provide some mercy in the form of an initial
+                    # hint for things we can't fully sniff yet
+                    initial_hints['encoding'] = test_details['initial']
+                out = sniff_hints_from_fileobjs(fileobjs=fileobjs,
+                                                initial_hints=initial_hints)
+            needed_settings = {
+                'compression': None,
+                'encoding': test_details['hint'],
+                'field-delimiter': ',',
+                'header-row': True,
+                'record-terminator': '\n'
+            }
+            self.assertTrue(set(needed_settings.items()).issubset(set(out.items())),
+                            f"Needed at least {needed_settings}, got {out}")
```

Comment:
I factored out the component-style tests out of a file in test/unit and left the unit-style tests there.